In [29]:
!pip install matplotlib google-cloud-storage

In [3]:
import collections
import copy
import hashlib
import io
import os
import subprocess
import textwrap
import time
import glob

from typing import List, Text

from PIL import Image

import cv2
import numpy as np
import pandas as pd
import tabulate
import seaborn as sns
from tqdm import tqdm
import ast


import tensorflow as tf

import matplotlib.pyplot as plt

In [4]:
from google.colab import auth

# Authenticate user for access. There will be a popup asking you to sign in with your user and approve access.
auth.authenticate_user()

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [89]:
#@title Global params

class Globals:
  # GCP project with GCS bucket of interest
  gcp_project = 'dx-scin-public' #@param

  # GCS bucket with data to read
  gcs_bucket_name = 'dx-scin-public-data' #@param

  # CSV of case metadata to read
  cases_csv = 'dataset/scin_cases.csv' #@param

  # CSV of label metadata to read
  labels_csv = 'dataset/scin_labels.csv' #@param

  # Images directory
  gcs_images_dir = 'dataset/images/' #@param

  ### Key column names
  image_path_columns = ['image_1_path', 'image_2_path', 'image_3_path']
  weighted_skin_condition_label = "weighted_skin_condition_label"
  skin_condition_label = "dermatologist_skin_condition_on_label_name"

  ###### Formed during execution:

  # Client for querying GCS
  gcs_storage_client = None

  # Bucket object for loading files
  gcs_bucket = None

  # pd.DataFrame for the loaded metadata_csv
  cases_df = None

  # pd.DataFrame for the loaded labels_csv
  cases_and_labels_df = None

print(f'GCS bucket name: {Globals.gcs_bucket_name}')
print(f'cases_csv: {Globals.cases_csv}')
print(f'labels_csv: {Globals.labels_csv}')
print(f'images dir: {Globals.gcs_images_dir}')

GCS bucket name: dx-scin-public-data
cases_csv: dataset/scin_cases.csv
labels_csv: dataset/scin_labels.csv
images dir: dataset/images/


## Load dataset

In [90]:
#@title Create a dataframe that contains the metadata and condition labels

from google.cloud import storage

def list_blobs(storage_client, bucket_name):
  """Helper to list blobs in a bucket (useful for debugging)."""
  blobs = storage_client.list_blobs(bucket_name)
  for blob in blobs:
    print(blob)

def initialize_df_with_metadata(bucket, csv_path):
  """Loads the given CSV into a pd.DataFrame."""
  df = pd.read_csv(io.BytesIO(bucket.blob(csv_path).download_as_string()), dtype={'case_id': str})
  df['case_id'] = df['case_id'].astype(str)
  return df

def augment_metadata_with_labels(df, bucket, csv_path):
  """Loads the given CSV into a pd.DataFrame."""
  labels_df = pd.read_csv(io.BytesIO(bucket.blob(csv_path).download_as_string()), dtype={'case_id': str})
  labels_df['case_id'] = labels_df['case_id'].astype(str)
  merged_df = pd.merge(df, labels_df, on='case_id')
  return merged_df

Globals.gcs_storage_client = storage.Client(Globals.gcp_project)
Globals.gcs_bucket = Globals.gcs_storage_client.bucket(
    Globals.gcs_bucket_name
)
Globals.cases_df = initialize_df_with_metadata(Globals.gcs_bucket, Globals.cases_csv)
Globals.cases_and_labels_df = augment_metadata_with_labels(Globals.cases_df, Globals.gcs_bucket, Globals.labels_csv)
print(len(Globals.cases_and_labels_df))

5033


In [91]:
Globals.cases_and_labels_df.columns

Index(['case_id', 'source', 'release', 'year', 'age_group', 'sex_at_birth',
       'fitzpatrick_skin_type',
       'race_ethnicity_american_indian_or_alaska_native',
       'race_ethnicity_asian', 'race_ethnicity_black_or_african_american',
       'race_ethnicity_hispanic_latino_or_spanish_origin',
       'race_ethnicity_middle_eastern_or_north_african',
       'race_ethnicity_native_hawaiian_or_pacific_islander',
       'race_ethnicity_white', 'race_ethnicity_other_race',
       'race_ethnicity_prefer_not_to_answer', 'textures_raised_or_bumpy',
       'textures_flat', 'textures_rough_or_flaky', 'textures_fluid_filled',
       'body_parts_head_or_neck', 'body_parts_arm', 'body_parts_palm',
       'body_parts_back_of_hand', 'body_parts_torso_front',
       'body_parts_torso_back', 'body_parts_genitalia_or_groin',
       'body_parts_buttocks', 'body_parts_leg', 'body_parts_foot_top_or_side',
       'body_parts_foot_sole', 'body_parts_other',
       'condition_symptoms_bothersome_appearan

In [92]:
Globals.cases_and_labels_df.sample(1)

case_id source release  year     age_group sex_at_birth  \
3825  5361080985667681926   SCIN   1.0.0  2023  AGE_60_TO_69       FEMALE   

     fitzpatrick_skin_type race_ethnicity_american_indian_or_alaska_native  \
3825                  FST3                                             NaN   

     race_ethnicity_asian race_ethnicity_black_or_african_american  ...  \
3825                  NaN                                      NaN  ...   

     dermatologist_gradable_for_fitzpatrick_skin_type_1  \
3825                                                YES   

     dermatologist_gradable_for_fitzpatrick_skin_type_2  \
3825                                                NaN   

     dermatologist_gradable_for_fitzpatrick_skin_type_3  \
3825                                                NaN   

     dermatologist_fitzpatrick_skin_type_label_1  \
3825                                        FST1   

     dermatologist_fitzpatrick_skin_type_label_2  \
3825                                         NaN   

     dermatologist_fitzpatrick_skin_type_label_3  \
3825                                         NaN   

     gradable_for_monk_skin_tone_india gradable_for_monk_skin_tone_us  \
3825                              True                           True   

     monk_skin_tone_label_india monk_skin_tone_label_us  
3825                        1.0                     2.0  

[1 rows x 73 columns]

## Important Functions

**Variables**
* df_original: The full, unmodified dataset containing all cases, metadata, and image paths. This is equivalent to Globals.cases_and_labels_df and can be used like a normal pandas DataFrame.
* df_filtered: A working copy of the dataset that you can safely modify, filter, or clean without affecting the original.
* image_dir : Use this directory to access images within google cloud


**Functions**
* read_image_from_gcs(gcs_path)
  - Downloads and decodes an image directly from your GCS bucket using the path stored in the dataset (e.g. "dataset/images/12345.png").
*  get_all_image_paths(df):
  - Extracts all unique image paths from the three image columns (image_1_path, image_2_path, image_3_path) in the dataset.
* show_case_images(case_id)
  - Displays all available images for a given case_id directly from GCS.
* convert_to_binary_var(col_name)
  - convert to binary values

  Note: add instructions if we want to analyze/change/decode all images together

In [93]:
df_original = Globals.cases_and_labels_df

In [94]:
df_filtered = df_original.copy()

In [95]:
image_dir = "/content/gcs_mount/dataset/images"

In [96]:
def read_image_from_gcs(gcs_path):
    """Reads and decodes an image from GCS (relative path)."""
    try:
        blob = Globals.gcs_bucket.blob(gcs_path)
        img_bytes = blob.download_as_bytes()
        img_array = np.frombuffer(img_bytes, np.uint8)
        img = cv2.imdecode(img_array, cv2.IMREAD_COLOR)
        if img is None:
            return None
        # Convert BGR (OpenCV) → RGB (matplotlib)
        img_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        return img_rgb
    except Exception as e:
        print(f"Failed to read {gcs_path}: {e}")
        return None


In [97]:
def get_all_image_paths(df):
    """
    Flatten all image_1/2/3 columns from the given DataFrame
    into a unique list of image paths.
    """
    image_cols = ['image_1_path', 'image_2_path', 'image_3_path']

    all_paths = (
        df[image_cols]
        .stack()          # combine all image columns vertically
        .dropna()         # remove missing entries
        .unique()         # only keep unique paths
        .tolist()         # convert to list
    )

    print(f"📸 Found {len(all_paths)} unique image paths.")
    return all_paths


In [98]:
def show_case_images(case_id):
    """Display all available images for a given case_id."""
    row = Globals.cases_and_labels_df[Globals.cases_and_labels_df["case_id"] == str(case_id)]
    if row.empty:
        print(f"No case found for ID {case_id}")
        return

    paths = row[['image_1_path', 'image_2_path', 'image_3_path']].dropna(axis=1).values.flatten()
    plt.figure(figsize=(15, 5))
    for i, path in enumerate(paths, 1):
        img = read_image_from_gcs(path)
        plt.subplot(1, len(paths), i)
        plt.imshow(img)
        plt.title(f"{os.path.basename(path)}", fontsize=9)
        plt.axis("off")
    plt.suptitle(f"Case ID: {case_id}", fontsize=12)
    plt.tight_layout()
    plt.show()


In [99]:
def convert_to_binary_var(col_name):
  df_filtered[col_name].replace('YES', 1, inplace = True)
  df_filtered[col_name].fillna(0, inplace = True)
  print(df_filtered[col_name].value_counts())
  print("")


## No longer need to run this -- Feature Cleaning, Preprocessing, and Exploration


Note: Out of 5033 cases, there are 143 problematic rows where the weight equals to above 1.0 with few cases of 0.99 (cross checked with original excel sheet)

In [100]:
# df_filtered['fitzpatrick_skin_type'].fillna('NONE_IDENTIFIED', inplace = True)
# print(df_filtered['fitzpatrick_skin_type='].count())
# df_filtered['fitzpatrick_skin_type'].value_counts()


5033


/tmp/ipython-input-3525839076.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_filtered['fitzpatrick_skin_type'].fillna('NONE_IDENTIFIED', inplace = True)


fitzpatrick_skin_type
NONE_IDENTIFIED    2858
FST3                661
FST2                542
FST4                427
FST5                214
FST1                188
FST6                143
Name: count, dtype: int64

In [101]:
'''
race_ethnicity_cols_list = ['race_ethnicity_american_indian_or_alaska_native', 'race_ethnicity_asian', 'race_ethnicity_black_or_african_american', 'race_ethnicity_hispanic_latino_or_spanish_origin', 'race_ethnicity_middle_eastern_or_north_african', 'race_ethnicity_native_hawaiian_or_pacific_islander', 'race_ethnicity_white', 'race_ethnicity_other_race', 'race_ethnicity_prefer_not_to_answer', 'race_ethnicity_two_or_more_after_mitigation']

for col in race_ethnicity_cols_list:
  convert_to_binary_var(col)
'''

race_ethnicity_american_indian_or_alaska_native
0.0    4960
1.0      73
Name: count, dtype: int64

race_ethnicity_asian
0.0    4937
1.0      96
Name: count, dtype: int64

race_ethnicity_black_or_african_american
0.0    4735
1.0     298
Name: count, dtype: int64

race_ethnicity_hispanic_latino_or_spanish_origin
0.0    4743
1.0     290
Name: count, dtype: int64

race_ethnicity_middle_eastern_or_north_african
0.0    5026
1.0       7
Name: count, dtype: int64

race_ethnicity_native_hawaiian_or_pacific_islander
0.0    5029
1.0       4
Name: count, dtype: int64

race_ethnicity_white
0.0    3160
1.0    1873
Name: count, dtype: int64

race_ethnicity_other_race
0.0    5017
1.0      16
Name: count, dtype: int64

race_ethnicity_prefer_not_to_answer
0.0    4999
1.0      34
Name: count, dtype: int64

race_ethnicity_two_or_more_after_mitigation
0.0    4950
1.0      83
Name: count, dtype: int64



/tmp/ipython-input-3518719989.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_filtered[col_name].replace('YES', 1, inplace = True)
/tmp/ipython-input-3518719989.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_filtered[col_name].replace('YES', 1, inplace = True)


In [102]:
'''
texture_cols_list = ['textures_raised_or_bumpy', 'textures_flat', 'textures_rough_or_flaky', 'textures_fluid_filled']

for col in texture_cols_list:
  convert_to_binary_var(col)
'''

textures_raised_or_bumpy
1.0    2915
0.0    2118
Name: count, dtype: int64

textures_flat
0.0    4211
1.0     822
Name: count, dtype: int64

textures_rough_or_flaky
0.0    4001
1.0    1032
Name: count, dtype: int64

textures_fluid_filled
0.0    4404
1.0     629
Name: count, dtype: int64



/tmp/ipython-input-3518719989.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_filtered[col_name].replace('YES', 1, inplace = True)
/tmp/ipython-input-3518719989.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_filtered[col_name].replace('YES', 1, inplace = True)


In [103]:
'''
body_parts_col_start = df_original.columns.get_loc('body_parts_head_or_neck')
body_parts_col_end = df_original.columns.get_loc('body_parts_other')
body_parts_cols_list = df_filtered.iloc[:, body_parts_col_start:body_parts_col_end + 1 ]

for col in body_parts_cols_list:
  convert_to_binary_var(col)
'''

body_parts_head_or_neck
0.0    4296
1.0     737
Name: count, dtype: int64

body_parts_arm
0.0    3326
1.0    1707
Name: count, dtype: int64

body_parts_palm
0.0    4818
1.0     215
Name: count, dtype: int64

body_parts_back_of_hand
0.0    4512
1.0     521
Name: count, dtype: int64

body_parts_torso_front
0.0    4238
1.0     795
Name: count, dtype: int64

body_parts_torso_back
0.0    4381
1.0     652
Name: count, dtype: int64

body_parts_genitalia_or_groin
0.0    4811
1.0     222
Name: count, dtype: int64

body_parts_buttocks
0.0    4642
1.0     391
Name: count, dtype: int64

body_parts_leg
0.0    3339
1.0    1694
Name: count, dtype: int64

body_parts_foot_top_or_side
0.0    4618
1.0     415
Name: count, dtype: int64

body_parts_foot_sole
0.0    4929
1.0     104
Name: count, dtype: int64

body_parts_other
0.0    4490
1.0     543
Name: count, dtype: int64



/tmp/ipython-input-3518719989.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_filtered[col_name].replace('YES', 1, inplace = True)
/tmp/ipython-input-3518719989.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_filtered[col_name].replace('YES', 1, inplace = True)


In [104]:
'''
symptoms_col_start = df_original.columns.get_loc('condition_symptoms_bothersome_appearance')
symptoms_col_end = df_original.columns.get_loc('other_symptoms_no_relevant_symptoms')
symptoms_cols_list = df_filtered.iloc[:, symptoms_col_start:symptoms_col_end + 1 ]

for col in symptoms_cols_list:
  convert_to_binary_var(col)
'''

condition_symptoms_bothersome_appearance
0.0    3503
1.0    1530
Name: count, dtype: int64

condition_symptoms_bleeding
0.0    4808
1.0     225
Name: count, dtype: int64

condition_symptoms_increasing_size
0.0    4020
1.0    1013
Name: count, dtype: int64

condition_symptoms_darkening
0.0    4648
1.0     385
Name: count, dtype: int64

condition_symptoms_itching
1.0    2712
0.0    2321
Name: count, dtype: int64

condition_symptoms_burning
0.0    4045
1.0     988
Name: count, dtype: int64

condition_symptoms_pain
0.0    4299
1.0     734
Name: count, dtype: int64

condition_symptoms_no_relevant_experience
0.0    4725
1.0     308
Name: count, dtype: int64

other_symptoms_fever
0.0    4943
1.0      90
Name: count, dtype: int64

other_symptoms_chills
0.0    4913
1.0     120
Name: count, dtype: int64

other_symptoms_fatigue
0.0    4618
1.0     415
Name: count, dtype: int64

other_symptoms_joint_pain
0.0    4733
1.0     300
Name: count, dtype: int64

other_symptoms_mouth_sores
0.0    4937
1.0 

/tmp/ipython-input-3518719989.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_filtered[col_name].replace('YES', 1, inplace = True)
/tmp/ipython-input-3518719989.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_filtered[col_name].replace('YES', 1, inplace = True)


In [105]:
'''
df_filtered['related_category'].fillna("NO_RESPONSE", inplace=True)
df_filtered['condition_duration'].fillna("NO_RESPONSE", inplace=True)
df_filtered['image_2_shot_type'].fillna("N/A", inplace=True)
df_filtered['image_3_shot_type'].fillna("N/A", inplace=True)
df_filtered['combined_race'].fillna("NO_RESPONSE", inplace=True)
'''

/tmp/ipython-input-3035444516.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_filtered['related_category'].fillna("NO_RESPONSE", inplace=True)
/tmp/ipython-input-3035444516.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inp

In [106]:
'''
df_filtered = df_filtered[
    ~(
        (df_filtered['dermatologist_skin_condition_on_label_name'].isna() | (df_filtered['dermatologist_skin_condition_on_label_name'].astype(str) == '[]')) &
        (df_filtered['dermatologist_skin_condition_confidence'].isna() | (df_filtered['dermatologist_skin_condition_confidence'].astype(str) == '[]'))
    )
]
print(len(df_filtered))
'''

3061


In [107]:
'''
print(df_filtered['dermatologist_skin_condition_on_label_name'].value_counts()[:30])
print(df_filtered['dermatologist_skin_condition_confidence'].value_counts()[:30])
'''

dermatologist_skin_condition_on_label_name
['Eczema']                                                                  120
['Urticaria']                                                                81
['Eczema', 'Allergic Contact Dermatitis']                                    67
['Allergic Contact Dermatitis', 'Irritant Contact Dermatitis']               60
['Allergic Contact Dermatitis']                                              42
['Folliculitis']                                                             37
['Urticaria', 'Insect Bite', 'Allergic Contact Dermatitis']                  36
['Acute dermatitis, NOS']                                                    27
['Insect Bite']                                                              27
['Tinea', 'Psoriasis', 'Eczema']                                             23
['Urticaria', 'Allergic Contact Dermatitis']                                 22
['Psoriasis', 'Eczema']                                                      

In [108]:
'''

# Create a temporary DataFrame for this check, working with df_original
temp_df = df_original.copy()

# Convert stringified dictionaries in 'weighted_skin_condition_label' to actual dicts
temp_df['parsed_weights'] = temp_df['weighted_skin_condition_label'].apply(
    lambda x: ast.literal_eval(x) if isinstance(x, str) else x
)

# Calculate the sum of weights for each case_id from the parsed dictionary
def sum_dict_values(weight_dict):
    if isinstance(weight_dict, dict):
        # Sum only numeric values, ignore None if present in dictionary values
        return sum(v for v in weight_dict.values() if isinstance(v, (int, float)))
    return np.nan # Return NaN if not a dictionary or empty

temp_df['original_weights_sum'] = temp_df['parsed_weights'].apply(sum_dict_values)

# Identify cases where the sum is not approximately 1 (ignoring NaNs)
original_problematic_cases = temp_df[
    ~np.isclose(temp_df['original_weights_sum'], 1.0, equal_nan=True)
]

if original_problematic_cases.empty:
    print("All original 'weighted_skin_condition_label' dictionaries sum approximately to 1.")
else:
    print("Case IDs where original 'weighted_skin_condition_label' sums do NOT approximate 1:")
    # Display case_id, original dictionary, and its sum
    display(original_problematic_cases[['case_id', 'weighted_skin_condition_label', 'original_weights_sum']])
'''

Case IDs where original 'weighted_skin_condition_label' sums do NOT approximate 1:


case_id                      weighted_skin_condition_label  \
2     -1003358831658393077  {'Impetigo': 0.55, 'Herpes Zoster': 0.23, 'Bul...   
3     -1003826561155964328                                                 {}   
4     -1003844406100696311  {'Lichen planus/lichenoid eruption': 0.33, 'Fo...   
10    -1031911667791094656  {'Eczema': 0.55, 'Drug Rash': 0.23, 'Allergic ...   
12    -1036016816671158395                                                 {}   
...                    ...                                                ...   
5025    963694477603925530  {'Insect Bite': 0.55, 'Urticaria': 0.23, 'Alle...   
5026    964881286184961934                                                 {}   
5027    966663640331506400  {'Acne': 0.33, 'Molluscum Contagiosum': 0.33, ...   
5029    973872028150293734                                                 {}   
5031    995820220956352730  {'Erythema gyratum repens': 0.33, 'Seborrheic ...   

      original_weights_sum  
2                     1.01  
3                     0.00  
4                     0.99  
10                    1.01  
12                    0.00  
...                    ...  
5025                  1.01  
5026                  0.00  
5027                  0.99  
5029                  0.00  
5031                  0.99  

[2879 rows x 3 columns]

In [109]:
# Assuming 'problematic_cases' from the previous output is still available
'''
if not problematic_cases.empty:
    print("Counts of each problematic sum value:")
    display(problematic_cases.value_counts().sort_index())
else:
    print("No problematic cases to summarize.")
'''

Counts of each problematic sum value:


condition_weight
0.99    377
0.99      2
1.01    402
1.02      2
1.03      1
       ... 
2.72      7
2.75      1
2.82      1
2.86      1
3.00     13
Name: count, Length: 143, dtype: int64

In [110]:
# Assuming 'problematic_cases' is available from the previous step
'''
if not problematic_cases.empty:
    # Filter problematic_cases for sums greater than 1.01
    cases_gt_1_01 = problematic_cases[problematic_cases > 1.01]

    print(f"Number of cases where condition_weight sum is greater than 1.01: {len(cases_gt_1_01)}")
    if not cases_gt_1_01.empty:
        print("\nCounts of each sum value greater than 1.01:")
        display(cases_gt_1_01.value_counts().sort_index())
else:
    print("No problematic cases found.")
'''

Number of cases where condition_weight sum is greater than 1.01: 389

Counts of each sum value greater than 1.01:


condition_weight
1.02     2
1.03     1
1.17     2
1.18     1
1.19     1
        ..
2.72     7
2.75     1
2.82     1
2.86     1
3.00    13
Name: count, Length: 140, dtype: int64

Out of 5033 cases, there are 143 problematic rows where the weight equals to above 1.0

In [111]:
'''
# Convert stringified lists/dicts to real Python objects
for col in [
    "dermatologist_skin_condition_on_label_name",
    "dermatologist_skin_condition_confidence",
    "weighted_skin_condition_label",
]:
    df_filtered[col] = df_filtered[col].apply(
        lambda x: ast.literal_eval(x) if isinstance(x, str) else x
    )

# Create a list of weights from the dictionary, aligned with condition names, before exploding
def align_weights(row):
    conditions = row["dermatologist_skin_condition_on_label_name"]
    weights_dict = row["weighted_skin_condition_label"]
    if isinstance(conditions, list) and isinstance(weights_dict, dict):
        return [weights_dict.get(cond) for cond in conditions]
    return [None] * len(conditions) if isinstance(conditions, list) else None # Handle cases where conditions list exists but weights_dict is missing

df_filtered["aligned_condition_weights"] = df_filtered.apply(align_weights, axis=1)

# Now explode all three aligned lists
df_filtered = df_filtered.explode(
    [
        "dermatologist_skin_condition_on_label_name",
        "dermatologist_skin_condition_confidence",
        "aligned_condition_weights",  # Add the new aligned weights column here
    ]
).reset_index(drop=True)

# Rename for clarity
df_filtered = df_filtered.rename(columns={
    "dermatologist_skin_condition_on_label_name": "condition_name",
    "dermatologist_skin_condition_confidence": "condition_confidence",
    "aligned_condition_weights": "condition_weight"  # Rename to the desired final column name
})

# Drop the original dictionary column for weights as it's no longer needed in its original form
df_filtered = df_filtered.drop(columns=["weighted_skin_condition_label"])

# Preview final structure
print(df_filtered[["case_id", "condition_name", "condition_confidence", "condition_weight"]].head(10))
'''

                case_id                    condition_name  \
0  -1000600354148496558            Inflicted skin lesions   
1  -1000600354148496558                            Eczema   
2  -1000600354148496558       Irritant Contact Dermatitis   
3  -1002039107727665188                 Prurigo nodularis   
4  -1002039107727665188                 Actinic Keratosis   
5  -1002039107727665188                         SCC/SCCIS   
6  -1003358831658393077                          Impetigo   
7  -1003358831658393077                     Herpes Zoster   
8  -1003358831658393077                Bullous dermatitis   
9  -1003844406100696311  Lichen planus/lichenoid eruption   

  condition_confidence condition_weight  
0                    4             0.41  
1                    4             0.41  
2                    3             0.18  
3                    4             0.41  
4                    3             0.18  
5                    4             0.41  
6                    2            

In [112]:
'''
import os

save_dir = "/content/drive/My Drive/BTT_Skinterest_2A/Dataset"
save_path = os.path.join(save_dir, "dermatologist_conditions_split.csv")
df_filtered.to_csv(save_path, index=False)
'''

 ## Image Cleaning

In [113]:
from google.cloud import storage
# List all images in dataset/images/
all_images = []
for blob in Globals.gcs_bucket.list_blobs(prefix=Globals.gcs_images_dir):
    if blob.name.endswith((".jpg",".jpeg",".png")):
        all_images.append(blob.name)

In [114]:
# Problematic images identified by CleanVision
to_drop = [
    "dataset/images/-4593817128438983108.png",
    "dataset/images/-2431769699504014881.png"
]

# Columns that contain image paths
image_cols = ['image_1_path', 'image_2_path', 'image_3_path']

# Build a boolean mask for rows that have any of these images
mask = df_filtered[image_cols].isin(to_drop).any(axis=1)

# Apply filter (keep only rows without those images)
df_filtered = df_filtered[~mask].copy()

In [115]:
save_dir = "/content/drive/My Drive/BTT_Skinterest_2A/Dataset"
os.makedirs(save_dir, exist_ok=True)

In [116]:
save_path = os.path.join(save_dir, "image_quality_report.csv")
quality_df = pd.read_csv(save_path)
print("Loaded", len(quality_df), "rows")
quality_df.head()

Loaded 10379 rows


image_path        blur  brightness_mean  \
0  dataset/images/-1001492676369731180.png   49.800900       125.822289   
1  dataset/images/-1001733364362669777.png  238.390468        74.623044   
2  dataset/images/-1003800477193786941.png   89.631693       128.779480   
3  dataset/images/-1005922060850163675.png    4.172010       102.138025   
4  dataset/images/-1007969568196430462.png    9.460231       147.597770   

   brightness_std  underexp   overexp  contrast    shadow  
0       62.991703  0.152014  0.049442  0.988142  0.162389  
1       29.900723  0.036180  0.000554  1.000000  0.036180  
2       47.623139  0.005774  0.023419  0.980237  0.055794  
3       62.690025  0.179942  0.000000  0.990610  0.010904  
4       40.054561  0.000000  0.002477  0.819608  0.107458

In [117]:
# Problematic images identified by CleanVision
to_drop_2 = [
    "dataset/images/4207723573736028617.png",
]

mask_2 = df_filtered[image_cols].isin(to_drop_2).any(axis=1)
df_filtered = df_filtered.loc[~mask_2]

print(f"Removed {mask_2.sum()} rows containing problematic images.")
print(f" Filtered dataset now has {len(df_filtered)} rows.")

Removed 0 rows containing problematic images.
 Filtered dataset now has 8209 rows.


In [118]:
save_path = os.path.join(save_dir, "image_quality_report_with_sharpness.csv")
quality_df = pd.read_csv(save_path)
print("Loaded", len(quality_df), "rows")
quality_df.head()

Loaded 8209 rows


/tmp/ipython-input-2692060160.py:2: DtypeWarning: Columns (69,70) have mixed types. Specify dtype option on import or set low_memory=False.
  quality_df = pd.read_csv(save_path)


case_id source release  year    age_group  \
0 -1000600354148496558   SCIN   1.0.0  2023  AGE_UNKNOWN   
1 -1000600354148496558   SCIN   1.0.0  2023  AGE_UNKNOWN   
2 -1000600354148496558   SCIN   1.0.0  2023  AGE_UNKNOWN   
3 -1002039107727665188   SCIN   1.0.0  2023  AGE_UNKNOWN   
4 -1002039107727665188   SCIN   1.0.0  2023  AGE_UNKNOWN   

           sex_at_birth fitzpatrick_skin_type  \
0  OTHER_OR_UNSPECIFIED       NONE_IDENTIFIED   
1  OTHER_OR_UNSPECIFIED       NONE_IDENTIFIED   
2  OTHER_OR_UNSPECIFIED       NONE_IDENTIFIED   
3  OTHER_OR_UNSPECIFIED       NONE_IDENTIFIED   
4  OTHER_OR_UNSPECIFIED       NONE_IDENTIFIED   

   race_ethnicity_american_indian_or_alaska_native  race_ethnicity_asian  \
0                                              0.0                   0.0   
1                                              0.0                   0.0   
2                                              0.0                   0.0   
3                                              0.0                   0.0   
4                                              0.0                   0.0   

   race_ethnicity_black_or_african_american  ...  \
0                                       0.0  ...   
1                                       0.0  ...   
2                                       0.0  ...   
3                                       0.0  ...   
4                                       0.0  ...   

   dermatologist_gradable_for_fitzpatrick_skin_type_1  \
0                                                YES    
1                                                YES    
2                                                YES    
3                                                YES    
4                                                YES    

   dermatologist_gradable_for_fitzpatrick_skin_type_2  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    

   dermatologist_gradable_for_fitzpatrick_skin_type_3  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    

   dermatologist_fitzpatrick_skin_type_label_1  \
0                                         FST2   
1                                         FST2   
2                                         FST2   
3                                         FST1   
4                                         FST1   

   dermatologist_fitzpatrick_skin_type_label_2  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

   dermatologist_fitzpatrick_skin_type_label_3  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

   gradable_for_monk_skin_tone_india  gradable_for_monk_skin_tone_us  \
0                               True                            True   
1                               True                            True   
2                               True                            True   
3                               True                            True   
4                               True                            True   

   monk_skin_tone_label_india  monk_skin_tone_label_us  
0                         2.0                      1.0  
1                         2.0                      1.0  
2                         2.0                      1.

# Training Dataset defined


In [59]:
df_filtered_path = "/content/drive/My Drive/BTT_Skinterest_2A/Dataset/dermatologist_conditions_split.csv"
df_filtered = pd.read_csv(df_filtered_path)
df_filtered[:10]

/tmp/ipython-input-932086250.py:2: DtypeWarning: Columns (68,69) have mixed types. Specify dtype option on import or set low_memory=False.
  df_filtered = pd.read_csv(df_filtered_path)


case_id source release  year     age_group  \
0 -1000600354148496558   SCIN   1.0.0  2023   AGE_UNKNOWN   
1 -1000600354148496558   SCIN   1.0.0  2023   AGE_UNKNOWN   
2 -1000600354148496558   SCIN   1.0.0  2023   AGE_UNKNOWN   
3 -1002039107727665188   SCIN   1.0.0  2023   AGE_UNKNOWN   
4 -1002039107727665188   SCIN   1.0.0  2023   AGE_UNKNOWN   
5 -1002039107727665188   SCIN   1.0.0  2023   AGE_UNKNOWN   
6 -1003358831658393077   SCIN   1.0.0  2023  AGE_18_TO_29   
7 -1003358831658393077   SCIN   1.0.0  2023  AGE_18_TO_29   
8 -1003358831658393077   SCIN   1.0.0  2023  AGE_18_TO_29   
9 -1003844406100696311   SCIN   1.0.0  2023  AGE_40_TO_49   

           sex_at_birth fitzpatrick_skin_type  \
0  OTHER_OR_UNSPECIFIED       NONE_IDENTIFIED   
1  OTHER_OR_UNSPECIFIED       NONE_IDENTIFIED   
2  OTHER_OR_UNSPECIFIED       NONE_IDENTIFIED   
3  OTHER_OR_UNSPECIFIED       NONE_IDENTIFIED   
4  OTHER_OR_UNSPECIFIED       NONE_IDENTIFIED   
5  OTHER_OR_UNSPECIFIED       NONE_IDENTIFIED   
6                  MALE       NONE_IDENTIFIED   
7                  MALE       NONE_IDENTIFIED   
8                  MALE       NONE_IDENTIFIED   
9                FEMALE                  FST3   

   race_ethnicity_american_indian_or_alaska_native  race_ethnicity_asian  \
0                                              0.0                   0.0   
1                                              0.0                   0.0   
2                                              0.0                   0.0   
3                                              0.0                   0.0   
4                                              0.0                   0.0   
5                                              0.0                   0.0   
6                                              0.0                   0.0   
7                                              0.0                   0.0   
8                                              0.0                   0.0   
9                                              0.0                   0.0   

   race_ethnicity_black_or_african_american  ...  \
0                                       0.0  ...   
1                                       0.0  ...   
2                                       0.0  ...   
3                                       0.0  ...   
4                                       0.0  ...   
5                                       0.0  ...   
6                                       0.0  ...   
7                                       0.0  ...   
8                                       0.0  ...   
9                                       0.0  ...   

   dermatologist_gradable_for_fitzpatrick_skin_type_2  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    
5                                                NaN    
6                                                NaN    
7                                                NaN    
8                                                NaN    
9                                                NaN    

   dermatologist_gradable_for_fitzpatrick_skin_type_3  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    
5                                                NaN    
6                                                NaN    
7                                                NaN    
8                                                NaN    
9                                                NaN    

   dermatologist_fitzpatrick_skin_type_label_1  \
0                                         FST2   
1                                    

## Baseline Model

In [119]:
import tensorflow as tf

from tensorflow.keras import datasets, layers, models
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split


In [120]:
# Get unique case IDs
case_ids = df_filtered["case_id"].unique()
# Split train and test cases
train_ids, test_ids = train_test_split(case_ids, test_size=0.4, random_state=42)

In [121]:
train_df = df_filtered[df_filtered["case_id"].isin(train_ids)]
test_df = df_filtered[df_filtered["case_id"].isin(test_ids)]

In [122]:
df_train_image_paths = get_all_image_paths(train_df)
df_test_image_paths = get_all_image_paths(train_df)

📸 Found 3869 unique image paths.
📸 Found 3869 unique image paths.
